Update patch for Swing-Material: MaterialFrame and Material Icons!

DragShot opened the pull request:
This patch includes the following changes:

- `MaterialTextField` and `MaterialPasswordField` can now have an accent color set for the focus line and floating label.
- The `MaterialFormatedTextField` behaves more like a common text field now, including a hint, a focus line and a floating label with the corresponding accent color.
- `MaterialComboBoxes` behave a bit more like text fields now, including an adjustable accent color for the focus line and arrow.
- The scrollbar used in combo boxes has now rounded bounds.
- Added several methods for convenient handling of windows.
- Added a cache layer for material shadows, so the CPU use decreases when the GUI is idle. As result, instances of `MaterialShadow` are now used per component.
- `MaterialButton` and `MaterialPane` can now adjust the foreground color based on the background color set.
- Imported some utilitary methods from [DS Desktop Notify](https://github.com/DragShot/DS-Desktop-Notify) to provide support with current, recently added and future features.
- Some docs have been added and/or extended.

Identified bugs and issues:

- In Windows environments with on-board Intel Graphics, the popup of a combo box might not display its elements as soon as it appears, being filled with the L&F default background color instead until the mouse enters the area. This doesn't seem to happen in other enviroments, like PCs with dedicated graphics or AMD integrated graphics.
- Manual element editing is not properly supported in `MaterialComboBoxes`.
- Some subclasses might need to be exported to their own file as top-level classes, as they are used by different components instead of only one.

Commit:
Minor fixes and some improvements were made:
- MaterialTextField and MaterialPasswordField can now have an accent color set for the focus line and floating label.
- The MaterialFormatedTextField behaves more like a common text field now, including a hint, a focus line and a floating label with the corresponding accent color.
- MaterialComboBoxes behave a bit more like text fields now, including an adjustable accent color for the focus line and arrow.
- The scrollbar used in combo boxes has now rounded bounds.
- Added several methods for convenient handling of windows.
- Added a cache layer for material shadows, so the CPU use decreases when the GUI is idle.
- MaterialButton and MaterialPane can now adjust the foreground color based on the background color set.
- Imported some utilitary methods from DS Desktop Unify to provide support with current, recently added and future features.
Identified bugs and issues:
- In Windows environments with on-board Intel Graphics, the popup of a combo box might not display its elements as soon as it appears, being filled with the L&F default background color instead until the mouse enters the area. This doesn't seem to happen in other enviroments, like PCs with dedicated graphics or AMD integrated graphics.
- Manual element editing is not properly supported in MaterialComboBoxes.
- Some subclasses might need to be exported to their own file as top-level classes, as they are used by different components instead of only one.

Commit:
Removed description inliners in preparation for first PR.

DragShot committed:
Added some methods in Utils for handling of colors.

DragShot committed:
Fixed RippleEffect not using the set rippleColor.

leMaik commented:
"Maintenance patch" :smile: I'd say that this is a pretty great feature patch, thank you very much! :+1:

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialPanel.java:
```diff
@@ -5,36 +5,80 @@
     /**
-     * Creates a new panel.
+     * Creates a new {@code MaterialPanel}. For the moment, this is the only
+     * supported constructor from JPanel.
```
It is, obviously, because it's the only constructor _available_ here.

DragShot replied:
_Until now_. If we manage to add support for `LayoutManager`s in the future, maybe adding a constructor for it in a similar fashion than JPanel wouldn't be a bad idea.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialPanel.java:
```diff
@@ -5,36 +5,80 @@
+     * supported constructor from JPanel.
      */
     public MaterialPanel() {
+        super();
```
There is no need to explicitly call the default `super()` constructor here.

DragShot replied:
Sorry, it's more like an habit I have. I am aware the compiler adds the call by itself, yet I always include it explicitly in source code. It helps me guiding myself when reading codes.

But this is your library, so...

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialPanel.java:
```diff
@@ -5,36 +5,80 @@
+     * take a little while for the resulting shadow to reflect the level once
+     * it is set.
      *
      * @param elevation elevation level (1..5)
```
Whoops, I documented the param wrong here, the elevation level is of course 0..5, as above. Would you like to fix that, @DragShot?

DragShot replied:
Alright.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialPanel.java:
```diff
@@ -5,36 +5,80 @@
+     * depending of how bright or dark is the chosen background color.
+     * <p/>
+     * <b>NOTE:</b> It is up to the look and feel to honor this property, some
+     * may choose to ignore it. To avoid any conflicts, using the Metal L&F is
```
Maybe it's because I'm not _that_ fluent in Swing, but I had to think a while to get what "Metal L&F" means. Could we change this to "Metal Look and Feel", perhabs with a link to the [corresponding javadoc](https://docs.oracle.com/javase/7/docs/api/javax/swing/plaf/metal/package-summary.html)?

DragShot replied:
Sure, I don't see why not.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,29 +8,216 @@
-        GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
-        if (ge.getDefaultScreenDevice().isWindowTranslucencySupported(GraphicsDevice.WindowTranslucency.PERPIXEL_TRANSLUCENT)) {
-            setUndecorated(true);
+        super();
```
Still no need to explicitly call this.

DragShot replied:
_See my reply above_.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,29 +8,216 @@
+     * just use it alone to switch the foreground/background colors.
+     * @return this window
+     */
+    public MaterialWindow dark() {
```
I'm not sure about these `dark()` and `light()` methods. Maybe we could add a `MaterialWindow.Style` enum with `LIGHT` and `DARK` values and add a `setStyle` method (a `getStyle` method would not really be needed for now). We could also add a constructor that takes a `style`.
What do you think?

DragShot replied:
Those are still experimental so I'm not touching them _just now_. I'm adding a `MaterialFrame` component extending from `MaterialWindow` in the next commit, so I want to see how it all fits first. They might be modified, moved into `MaterialPanel` or just dropped depending of the outcome.

DragShot replied:
Also, keep on mind that those two methods do not define styles at all. They are just shortcuts for a pair of `setForeground()` and `setBackground()` calls, although more specific functionality may be added in the future.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,29 +8,216 @@
+    
+    /**
+     * Alternative method to {@link JFrame#setLocation(int, int)}. Use it to set
+     * a new location on screen for this window, this method takes into
```
Maybe we should override `JFrame#setLocation`, just call `super.setLocation` and deprecate it and tell the user to use this instead to guide users of this to using this method.

DragShot replied:
Hmm... not a bad idea. I'm not a fan of deprecating API normally available as current from the library being extended, but at least a warning should be in place.

leMaik replied:
I'd just deprecate it in order to warn the user, so that the users explicitly needs to mute the warning.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,29 +8,216 @@
+     * consideration the shadow offsets so you don't have to worry about it.
+     * @param p the point defining the top-left corner of the new location in the desktop
+     */
+    public void setWindowLocation(Point p) {
```
This should just call `this.setWindowLocation(p.x, p.y);` instead of re-implementing that logic.

DragShot replied:
Alright, I'm changing that.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,29 +8,216 @@
+     * the shadow.
+     * @return the current x coordinate of this window's origin
+     */
+    public int getXLocation() {
```
I'd also suggest deprecating `JFrame#getX()`, in the same way as above.

DragShot replied:
_See my reply above_.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,29 +8,216 @@
+     * @return the current y coordinate of this window's origin
+     */
+    public int getYLocation() {
+        return super.getY() + (shadowPane == null ? 0:MaterialShadow.OFFSET_TOP);
```
I'd also suggest deprecating `JFrame#getY()`, in the same way as above.

DragShot replied:
_See my reply above_.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/Utils.java:
```diff
@@ -0,0 +1,158 @@
+import java.util.Arrays;
+
+/**
+ * Put miscellaneous utilitary static methods here.
```
This comment should not document what _we_ should do when we extend this library, but rather what this class does, e.g. "Various utility functions for swing-material."

DragShot replied:
Aww I thought you would get the reference.
Anyway, that issue shall be addressed.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/Utils.java:
```diff
@@ -0,0 +1,158 @@
+     *  1: Class/method is available.<br/>
+     * -1: Class/method is not available.
+     */
+    private static int useSun2D = 0;
```
Meh, integer flags. This should be a `private static final boolean useSun2D` and we can check statically if it is available:

```java
private static final boolean useSun2D;
static {
  useSun2D = ...; // check
}
```

DragShot replied:
I'm more a fan of lazy loading in these scenarios, but whatever works.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/Utils.java:
```diff
@@ -0,0 +1,158 @@
+    }
+    
+    //Uncomment this block in order to test #isDark() against all the color constants in Material Color
+    /*public static void main(String[] args) {
```
This could be added to an examples folder. :+1:

DragShot replied:
I'll try to come up with an example once the `MaterialFrame` is ready.

leMaik replied:
:+1: Then this code should be removed.

leMaik reviewed (changes requested):
First, I really really like your additions. :tada: I've got some small suggestions, though. But all in all, it's fine.

If you want, you can add yourself as `@author` to the classes you modified. I didn't add myself to every method, but if you want to keep that, it's ok.

DragShot committed:
Changes:
- Added new object MaterialFrame, for decorated windows that resemble classic frames. Includes a colorizable title bar, control buttons and the ability of dragging the frame around the screen, all out of the box.
- Added new object MaterialFrameWrapper. It handles the decorations and functionality associated with MaterialFrames.
- Several methods inherited from JFrame were deprecated in MaterialWindow, as better alternatives have been provided.
- Added a new font resource: Material Icons, from Google.
- Added new class MaterialIcons, providing access to the embbeded iconfont and to the char mappings for each icon available.
- Moved class Roboto into a separate fonts package.
- Some docs have been added.
- Other edits and minor bug fixes.

DragShot commented:
More stuff has been done:

- Added new object `MaterialFrame`, for decorated windows that resemble classic frames. Includes a colorizable title bar, control buttons and the ability of dragging the frame around the screen, all out of the box.
- Added new object `MaterialFrameWrapper`. It handles the decorations and functionality associated with `MaterialFrame`s.
- Several methods inherited from `JFrame` were deprecated in `MaterialWindow`, as better alternatives have been provided.
- Added a new font resource: Material Icons, from Google.
- Added new class `MaterialIcons`, providing access to the embbeded iconfont and to the char mappings for each icon available.
- Moved class `Roboto` into a separate fonts package.
- Some docs have been added.
- Other edits as requested, and minor bug fixes.

Now this is truly more an update than a patch :smiley: .

Here's some code I used to test the Frame:

```java
public static void main(String[] args) {
    try {
        javax.swing.UIManager.setLookAndFeel(new javax.swing.plaf.metal.MetalLookAndFeel());
    } catch (Exception ex) {
        ex.printStackTrace();
    }
    Color baseColor = MaterialColor.INDIGO_500;
    Color accentColor = MaterialColor.PINKA_400;
    MaterialFrame frame = new MaterialFrame();
    frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
    frame.setTitle("Swing Material");
    frame.setColor(baseColor);
    frame.setWindowSize(640, 480);
    MaterialButton button = new MaterialButton();
    button.setBackground(baseColor);
    button.setText("This is Button");
    frame.add(button);
    button.setBounds(-MaterialShadow.OFFSET_LEFT,
        -MaterialShadow.OFFSET_TOP,
        210+MaterialShadow.OFFSET_LEFT+MaterialShadow.OFFSET_RIGHT,
        41+MaterialShadow.OFFSET_TOP+MaterialShadow.OFFSET_BOTTOM);
    button.setLocation(frame.getContentPane().getWidth()-button.getWidth(),
        frame.getContentPane().getHeight()-button.getHeight());

    MaterialPasswordField passField = new MaterialPasswordField();
    passField.setLabel("Input password field");
    passField.setHint("Put password here");
    passField.setSize(240, 72);
    passField.setLocation(MaterialShadow.OFFSET_LEFT,
        frame.getContentPane().getHeight() - passField.getHeight() - MaterialShadow.OFFSET_BOTTOM/2);
    passField.setAccent(accentColor);
    frame.add(passField);

    MaterialTextField textField = new MaterialTextField();
    textField.setLabel("Input text field");
    textField.setHint("Put text here");
    textField.setSize(passField.getWidth(), passField.getHeight());
    textField.setLocation(MaterialShadow.OFFSET_LEFT,
        frame.getContentPane().getHeight() - textField.getHeight()*2 - MaterialShadow.OFFSET_BOTTOM/2);
    textField.setAccent(accentColor);
    frame.add(textField);

    //frame.dark(); //Experimental
    //frame.setResizable(false); //Uncommenting this actually does something!
    frame.setLocationRelativeTo(null);
    frame.setVisible(true);
    System.out.println(frame.getWidth()+"x"+frame.getHeight()+" ("+frame.getWindowWidth()+"x"+frame.getWindowHeight()+")");
}
```

DragShot commented:
Hmm... it seems some files made it there when they shouldn't. I'll clean that out later. In the meantime, why don't you guys try the `MaterialFrame` and send some screenshots of it working?

leMaik commented:
@DragShot I'll try it out as soon as I can. If there are no breaking changes, I should be able to port our existing program and use the `MaterialFrame` for it.

DragShot committed:
Changes:
- Several methods inherited from JFrame were deprecated in MaterialWindow, as better alternatives have been provided.
- Removed leftover files from a backup.
- Some docs have been added.

DragShot committed:
Some docs have been added.

DragShot committed:
Some docs have been added.Changes:
- Now MaterialButtons can have a customized border radius. With this they can be rounded or even look like pills, and their shadows will follow that appearance.
- ShadowEffect now supports border radius in square shadows.
- Square MaterialShadows now supports specifying a border radius.
- MaterialTextField, MaterialPasswordField, MaterialFormattedTextField and MaterialComboBox now use the current foreground color in order to display unfocused labels and hints.
- Using a solid color in a MaterialPanel no longer hides the shadow painted by the ElevationEffect.
- Setting an elevation level in an ElevationEffect, if the target component isn't currently visible, will no longer trigger an animation.
- Some docs have been added.

DragShot committed:
Fixed MaterialShadow not considering the border radius in the cache.

DragShot commented:
I've added some more stuff I could need:
- Now `MaterialButton`s can have a customized border radius. With this they can be rounded or even look like pills, and their shadows will follow that appearance.
- `ShadowEffect` now supports border radius in square shadows.
- Square `MaterialShadow`s now supports specifying a border radius.
- `MaterialTextField`, `MaterialPasswordField`, `MaterialFormattedTextField` and `MaterialComboBox` now use the current foreground color in order to display unfocused labels and hints.
- Using a solid color in a `MaterialPanel` no longer hides the shadow painted by the `ElevationEffect`.
- Setting an elevation level in an `ElevationEffect`, if the target component isn't currently visible, will no longer trigger an animation.
- Some docs have been added.

DragShot commented:
Here's another test with the latest changes, this time as a small login window:

![LoginFrame](https://i58.servimg.com/u/f58/17/37/76/58/swing-10.png)

Code:

```java
package ds.gui.swing.material;

import de.craften.ui.swingmaterial.MaterialButton;
import de.craften.ui.swingmaterial.MaterialColor;
import de.craften.ui.swingmaterial.MaterialFrame;
import de.craften.ui.swingmaterial.MaterialPanel;
import de.craften.ui.swingmaterial.MaterialPasswordField;
import de.craften.ui.swingmaterial.MaterialShadow;
import de.craften.ui.swingmaterial.MaterialTextField;
import de.craften.ui.swingmaterial.fonts.MaterialIcons;
import java.awt.BorderLayout;
import java.awt.Color;
import javax.swing.JLabel;
import javax.swing.JPanel;

/**
 * A sample {@link MaterialFrame} demonstrating Swing Material.
 * @author DragShot
 */
public class LoginFrame extends MaterialFrame {
//    static final Color MAIN_COLOR = MaterialColor.INDIGO_500;
//    static final Color SECONDARY_COLOR = MaterialColor.INDIGO_700;
//    static final Color ACCENT_COLOR = MaterialColor.PINKA_400;
    static final Color MAIN_COLOR = MaterialColor.TEAL_600;
    static final Color SECONDARY_COLOR = MaterialColor.TEAL_800;
    static final Color ACCENT_COLOR = MaterialColor.AMBERA_400;
    
    MaterialPanel pnlLogin;
    MaterialTextField txtLogin;
    MaterialPasswordField pwdLogin;
    MaterialButton btnLogin;
    JLabel loginIcn;
    
    public LoginFrame() {
        initComponents();
        setTitle("Material Login");
        setColor(MAIN_COLOR);
        //dark();
        setLayout(null);
        setWindowSize(360, 580);
        setResizable(false);
        setLocationRelativeTo(null);
    }
    
    private void initComponents() {
        pnlLogin = new MaterialPanel();
        pnlLogin.setLayout(new BorderLayout());
        pnlLogin.setBackground(SECONDARY_COLOR);
        pnlLogin.setElevation(2);
        pnlLogin.setLayout(null);
        loginIcn = new JLabel();
        loginIcn.setFont(MaterialIcons.ICON_FONT.deriveFont(14f));
        loginIcn.setHorizontalAlignment(JLabel.CENTER);
        loginIcn.setVerticalAlignment(JLabel.BOTTOM);
        loginIcn.setText(String.valueOf(MaterialIcons.ACCOUNT_CIRCLE));
        loginIcn.setForeground(new Color(pnlLogin.getForeground().getRGB() & 0x00FFFFFF | 0xBB000000, true));
        pnlLogin.add(loginIcn);
        txtLogin = new MaterialTextField();
        txtLogin.setLabel("Account ID");
        txtLogin.setHint("name@example.com");
        txtLogin.setAccent(ACCENT_COLOR);
        txtLogin.setForeground(pnlLogin.getForeground());
        pnlLogin.add(txtLogin);
        pwdLogin = new MaterialPasswordField();
        pwdLogin.setLabel("Password");
        pwdLogin.setHint("A-Z a-z 1-9");
        pwdLogin.setAccent(ACCENT_COLOR);
        pwdLogin.setForeground(pnlLogin.getForeground());
        pnlLogin.add(pwdLogin);
        this.add(pnlLogin);
        btnLogin = new MaterialButton();
        btnLogin.setType(MaterialButton.Type.RAISED);
        btnLogin.setText("Sign in");
        btnLogin.setBackground(MAIN_COLOR);
        this.add(btnLogin);
    }
    
    @Override
    public void doLayout() {
        super.doLayout();
        int wd = getContentPane().getWidth();
        int ht = getContentPane().getHeight();
        btnLogin.setSize(wd - MaterialShadow.OFFSET_LEFT - MaterialShadow.OFFSET_RIGHT,
                41 + MaterialShadow.OFFSET_TOP + MaterialShadow.OFFSET_BOTTOM);
        btnLogin.setLocation(MaterialShadow.OFFSET_LEFT, ht - btnLogin.getHeight() - MaterialShadow.OFFSET_BOTTOM);
        btnLogin.setBorderRadius((btnLogin.getHeight() - (MaterialShadow.OFFSET_TOP + MaterialShadow.OFFSET_BOTTOM))/2);
        pnlLogin.setBounds(-MaterialShadow.OFFSET_LEFT, -MaterialShadow.OFFSET_TOP,
                wd + MaterialShadow.OFFSET_LEFT + MaterialShadow.OFFSET_RIGHT,
                ht - btnLogin.getHeight() - MaterialShadow.OFFSET_BOTTOM/* + MaterialShadow.OFFSET_TOP*/);
        pwdLogin.setSize(pnlLogin.getWidth() - MaterialShadow.OFFSET_LEFT*3 - MaterialShadow.OFFSET_RIGHT*3, 72);
        pwdLogin.setLocation(MaterialShadow.OFFSET_LEFT*3, pnlLogin.getHeight() - MaterialShadow.OFFSET_TOP*2 - MaterialShadow.OFFSET_BOTTOM*2 - pwdLogin.getHeight() + 15);
        txtLogin.setSize(pwdLogin.getWidth(), pwdLogin.getHeight());
        txtLogin.setLocation(pwdLogin.getX(), pwdLogin.getY() - txtLogin.getHeight());
        loginIcn.setBounds(MaterialShadow.OFFSET_LEFT*2, MaterialShadow.OFFSET_TOP*2,
                pnlLogin.getWidth() - MaterialShadow.OFFSET_LEFT*2 - MaterialShadow.OFFSET_RIGHT*2,
                txtLogin.getY());
        loginIcn.setFont(loginIcn.getFont().deriveFont((float)Math.min(loginIcn.getWidth(), loginIcn.getHeight())));
    }
    
    public static void main(String[] args) {
        System.setProperty("awt.useSystemAAFontSettings","lcd");
        System.setProperty("swing.aatext", "true");
        try {
            javax.swing.UIManager.setLookAndFeel(new javax.swing.plaf.metal.MetalLookAndFeel());
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        java.awt.EventQueue.invokeLater(new Runnable() {@Override public void run() {
            LoginFrame frame = new LoginFrame();
            frame.setDefaultCloseOperation(javax.swing.JFrame.EXIT_ON_CLOSE);
            frame.setVisible(true);
        }});
    }
}
```

leMaik commented:
@DragShot Nice new additions! It would be cool to "freeze" the features of this PR to make reviewing it a little bit easier? I'd really like to merge this, but the awesome new features you add make that pretty hard. :sweat_smile:

pedrohidalgo commented:
I'm waiting for this pull request to be accepted. I'm planning to add some documentation while I'm learning how to use each component. But I think I have to wait for these new cool additions to be merged because they probably already changed some existing functionalities.

Very good work guys 👍

leMaik commented:
@pedrohidalgo Yeah, some documentation would be great! :+1: 
I'm pretty happy with the javadoc right now, but maybe we should add some github pages to host them online and provide some additional help there.

pedrohidalgo commented:
Good Idea. I was planning first to add a demo folder with a readme and some sample classes with a main method so they can be run quickly on any IDE.

The github pages could be a second step for me since I think they would take more time.

DragShot committed:
Minor adjustments in default border radius values.

DragShot committed:
Some docs have been modified.

DragShot commented:
Alright, I think that will be all for this pull request. Posterior changes won't be pushed here until this PR is accepted.

In the meantime I will be busy with the development of my own app using this library.

leMaik commented:
@DragShot Great, and thank you again for all the great changes! :tada:
I'll take a closer look at the code soon, but I think I'll just merge it and do small modifications after that.

DragShot commented:
Magnificent!

I guess we can all try to come up with examples while you're at it. And maybe a more elaborated readme.

leMaik commented:
> And maybe a more elaborated readme.

Something with more than one sentence would be great, indeed! :smile: 

Regarding the example: I'm looking forward to updating our [Minecraft launcher](https://github.com/TeamWertarbyte/craften-launcher) soon.

pedrohidalgo commented:
@DragShot I'm planning to start working with examples as soon as @leMaik Merge your changes, so I make sure I'm using the latest version of the code since your changes @DragShot were huge!.

btw, why are we using java 7 and not java 8 on this project. Any special need?

leMaik commented:
@pedrohidalgo Our launcher needed to support Java 7, and we created this library for that launcher. I don't see any advantage of using java 8 for this project, so let's just stick to java 7 for now.

pedrohidalgo commented:
Ok, I agree.

About the docs: @DragShot @leMaik I think would be useful to have a demo JFrame showcasing all the components. We need to coordinate the documentation work so we don't overlap each other. Can I safely start using the @DragShot code for the docs and samples or is it better to wait for the merge?

leMaik commented:
@pedrohidalgo I don't think there will be any breaking changes to this PR anymore. Feel free to start! :+1:

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialButton.java:
```diff
@@ -154,16 +197,16 @@ protected void paintComponent(Graphics g) {
         if (isEnabled()) {
             g2.setColor(getBackground());
-            g2.fill(new RoundRectangle2D.Float(0, 0, getWidth() - offset_lr, getHeight() - offset_td, 3, 3));
+            g2.fill(new RoundRectangle2D.Float(0, 0, getWidth() - offset_lr, getHeight() - offset_td, borderRadius*2, borderRadius*2));
```
Why `borderRadius * 2`? The last two parameters are already the radius, aren't they?

DragShot replied:
Because `RoundRectangle2D` receives arc lengths, that is the diameter. Multiply the radius * 2 and you get the diameter.

leMaik replied:
Allright. :+1:

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialButton.java:
```diff
@@ -179,7 +222,7 @@ protected void paintComponent(Graphics g) {
 
         if (isEnabled()) {
-            g2.setClip(new RoundRectangle2D.Float(0, 0, getWidth() - offset_lr, getHeight() - offset_td, 3, 3));
+            g2.setClip(new RoundRectangle2D.Float(0, 0, getWidth() - offset_lr, getHeight() - offset_td, Math.max(borderRadius*2 - 4, 0), Math.max(borderRadius*2 - 4, 0)));
```
Where does the magic `4` come from? This should at least be documented somewhere.

DragShot replied:
That's for the ripple to not be left behind the button borders. As only the buttons use them *this* way, I see no reason to use a constant. If this were to change though, a constant pool should be considered for other classes to reuse this (and maybe some other constants could be pulled out from their current classes then).

leMaik replied:
:+1:

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialColor.java:
```diff
@@ -279,6 +279,7 @@
     public static final Color GREY_600 = Color.decode("#757575");
     public static final Color GREY_700 = Color.decode("#616161");
     public static final Color GREY_800 = Color.decode("#424242");
+    public static final Color GREY_850 = Color.decode("#313131");
```
I think we should stick to the [Material Design color palette](https://material.io/guidelines/style/color.html#color-color-palette).

DragShot replied:
See the [Themes section](https://material.io/guidelines/style/color.html#color-themes) in the Material Design Color Guide. Dark UIs use a color between 900 and 800 for the default content area (Nº 3). `MaterialColor.GREY_850` is meant to enact as such color in order to avoid unnecessary hardcoding.

Now that I see, maybe `MaterialWindow.light()` should set `MaterialColor.GREY_50` as background instead of `MaterialColor.WHITE`.

leMaik replied:
Okay, let's keep `GREY_850` then. Do you want to change `MaterialWindow.light()` to use `GREY_50`?

DragShot replied:
Already did that. Expect one last commit today, after I get home from work.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialComboBox.java:
```diff
@@ -38,19 +98,23 @@ public void paint(Graphics g) {
 
         g.setFont(Roboto.REGULAR.deriveFont(16f));
-        g2.setColor(Color.BLACK);
+        g.setColor(getSelectedItem() == null ? new Color(getForeground().getRGB() & 0x00FFFFFF | HINT_OPACITY_MASK, true):getForeground());
```
What does this line do exactly? :sweat_smile:

DragShot replied:
See [this](https://stackoverflow.com/questions/17256644/how-does-the-bitwise-and-work-in-java) on bitwise operators and [this](http://www.cafeaulait.org/course/week2/43.html) on conditional assignment.

*Technically*, it sets a color depending on if `getSelectedItem()` returns `null`. If it is, a new `Color` is created with the RGB values of `getForeground()` and a constant Alpha value (`getRGB()` returns an ARGB int value, where the first 8 bits are dropped with `& 0x00FFFFFF` and then replaced with our own via `| HINT_OPACITY_MASK`). If it isn't, the returned value of `getForeground()` is used instead.

*Practically*, it uses a translucent version of the foreground color to display the hint if there's no item selected (thus, no content to display). If there is something selected, the foreground color is used directly in order to paint the content.

Gotta love those one-liners.

leMaik replied:
For the sake of maintainability, please add this as a comment! :sweat_smile:

DragShot replied:
I've decided to move it into a method in Utils, so it can be properly commented and shared between components.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialComboBox.java:
```diff
@@ -38,19 +98,23 @@ public void paint(Graphics g) {
         g2.fillRect(0, getHeight() - 9, getWidth(), 1);
 
-        g2.fillPolygon(new int[]{getWidth() - 5, getWidth() - 10, getWidth() - 15}, new int[]{getHeight() - 24, getHeight() - 19, getHeight() - 24}, 3);
+        if (isFocusOwner()) g2.setColor(accentColor);
```
I prefer multiline `if` blocks, with brackets even if they just contain a single statements. Makes future changes a little less error prone.

DragShot replied:
Hmp. I don't see why, but I fail to see why *not* in the same way. I'll add some brackets.

leMaik replied:
Matter of taste. :wink:

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialComboBox.java:
```diff
@@ -89,7 +153,7 @@ public void paint(Graphics g) {
-                g2.setColor(MaterialColor.PINK_500);
+                g2.setColor(comboBox.accentColor);
             } else {
                 g2.setColor(Color.BLACK);
```
Shouldn't this be the foreground color now? Not sure about it, though. :thinking:

DragShot replied:
```java
public static class FieldRenderer<T> extends JComponent implements ListCellRenderer<T> {
```

The line you mention is using the accent color for the selected item in the list (actually taking place of `MaterialColor.PINK_500`).

Although now I see the `FieldRenderer` is still kind of using a fixed set of colors instead of the `MaterialComboBox`'s foreground and background. I think this should be addressed...

leMaik replied:
Maybe open a new issue so that we don't forget about this?

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFormattedTextField.java:
```diff
@@ -15,222 +17,253 @@
+
+        if (!getHint().isEmpty() && getText().length() == 0 && (getLabel().isEmpty() || isFocusOwner()) && floatingLabel.isFloatingAbove()) {
+            g.setFont(Roboto.REGULAR.deriveFont(16f));
+            g2.setColor(new Color(getForeground().getRGB() & 0x00FFFFFF | HINT_OPACITY_MASK, true));
```
What does this line do? I think a comment might be helpful.

DragShot replied:
*See my reply above.*

I agree a comment on top of it might be necessary.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrame.java:
```diff
@@ -0,0 +1,79 @@
+ * buttons and the ability to drag, minimize and maximize the window as the user
+ * wishes. Specially recommended when you want your application to behave like
+ * common desktop apps. If you are looking for a behavior more like mobile apps,
+ * try {@code MaterialAppFrame}.
```
Instead of `@code`, `@link` should be used here to link to `MaterialAppFrame` in the docs.

DragShot replied:
Oh wait, what is *that* doing there? I though I had removed it in the 3rd commit.

I mean, there's no such a thing as a `MaterialAppFrame` *yet*. We would need a proper Navbar and Sidebar first.

leMaik replied:
:smile: Allright, then this can be removed.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrame.java:
```diff
@@ -0,0 +1,79 @@
+    
+    @Override
+    protected void processWindowEvent(WindowEvent e) {
+        super.processWindowEvent(e);
```
This method only calls through to `super`, so we don't need it, right?

DragShot replied:
Actually, we *do*. `MaterialFrameWrapper` uses it in order to submit a `WINDOW_CLOSING` event into the `MaterialFrame` (it goes all the way back into `JFrame`) for proper handling of the default close operation and `windowClosing()` listeners. By overriding `processWindowEvent()` I'm sharing it within the package.

At least, that's the idea. So far it works.

leMaik replied:
Fine. :)

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrameWrapper.java:
```diff
@@ -0,0 +1,372 @@
+ * @author DragShot
+ */
+//The current implementation is based on Baltimore for AWT, an experimental
+//project I worked on some time ago.
```
Do you want to keep this comment?

DragShot replied:
You know what?... it won't be necessary.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrameWrapper.java:
```diff
@@ -0,0 +1,372 @@
+        return titleBar.getBackground();
+    }
+    
+    public void updateTitle() {
```
This method is missing javadoc for a public method.

DragShot replied:
Thank you for the heads-up.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrameWrapper.java:
```diff
@@ -0,0 +1,372 @@
+            this.frame = frame;
+            control = new FrameControler(frame);
+            Font font = MaterialIcons.ICON_FONT.deriveFont(20f);
+            btnClose = new MaterialButton();
```
We should add a `MaterialIconButton` that has an `icon` field that is of type `MaterialIcons`. :thinking: I'll add that when this PR is merged.

DragShot replied:
Sure, go ahead. Maybe you could try adding that functionality directly into `MaterialIconButton`.

leMaik replied:
Exactly, it should set its font and the font size.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrameWrapper.java:
```diff
@@ -0,0 +1,372 @@
+            btnClose.setType(MaterialButton.Type.FLAT);
+            //btnClose.setText("X");
+            btnClose.setFont(font);
+            btnClose.setText(String.valueOf(MaterialIcons.CLOSE));
```
Now while I'm thinking about it: Maybe `MaterialIcons` should be renamed `MaterialIcon` to match the `MaterialColor` enum (e.g. no plural)?

DragShot replied:
`MaterialIcons` was named after the font, just like `Roboto`. Also, `MaterialIcon` might conflict with a future extension of `ImageIcon` or similar, if ever needed.

leMaik replied:
Seems legit.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrameWrapper.java:
```diff
@@ -0,0 +1,372 @@
+        public void restore() {
+            int state = frame.getExtendedState();
+            // Clear the maximized bits
+            state &= ~Frame.MAXIMIZED_BOTH;
```
Is there a more Java-like way to do this (e.g. no bit arithmetic)?

DragShot replied:
**Short answer:** This is.

**Long answer:** That's exactly what `FrameControler` is for, which code you're looking right now. Using bit arithmetic is just what the Swing docs suggest to do (at least they did last time I checked them out).

leMaik replied:
Okay, fine.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrameWrapper.java:
```diff
@@ -0,0 +1,372 @@
+        }
+
+        /**
+         * This method maximizes a frame. The iconified bit is not affected
```
> the iconified bit

What?

DragShot replied:
See `FrameControler.iconify()` and `FrameControler.deiconify()`. Also, `Frame.ICONIFIED` and `Frame.setExtendedState()`.

Long story short, when a window is "iconified", it is minimized into the task bar. When you pull it out from it you "deiconify" it (I know how messed up that sounds, it took a while to me to understand how Java refers to each state too).

leMaik replied:
:+1: Thank you for the explanation!

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialFrameWrapper.java:
```diff
@@ -0,0 +1,372 @@
+            state |= Frame.MAXIMIZED_BOTH;
+            // Maximize the frame
+            frame.setExtendedState(state);
+        }
```
Same here, can this be done without bit operations?

DragShot replied:
*See my reply above.*

leMaik replied:
_Looks at reply above…_ _Agrees with reply above._

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialTextField.java:
```diff
@@ -206,12 +272,12 @@ void update() {
+                    targetColor = new Color(target.getForeground().getRGB() & 0x00FFFFFF | HINT_OPACITY_MASK, true);//MaterialColor.MIN_BLACK;
                 } else {
-                    targetColor = MaterialColor.LIGHT_BLACK;
+                    targetColor = accentColor; //MaterialColor.LIGHT_BLACK;
```
Why was this changed from light black to the accent color?

DragShot replied:
See the `LoginFrame` example I posted above. Both the floating label and focus line use the accent color when focused. The label remains floating and keeps the accent color when there is content to display.

An alternative could be switching to the hint color if you do not wish the accent color to remain when the target component loses focus.

leMaik replied:
We should take a look into the Material Design specs, and I think that the accent color doesn't remain if the component loses focus there.

DragShot replied:
Yeah, I've noticed. I'm fixing that.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,45 +8,439 @@
+        setUndecorated(true);
+        if (Utils.isTranslucencySupported()) {
+            setBackground(MaterialColor.TRANSPARENT);
+            //This default behavior should NOT be set here!
```
Indeed. :smile: You can delete these two lines.

DragShot replied:
Alright then.

We still need to figure out what keeps the VM running after disposing of all windows though.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,45 +8,439 @@
+     * {@inheritDoc}
+     * <p/>
+     * <b>NOTE:</b> If this window is casting a shadow, calling this method
+     * might provide unnacurate results unless you take into account the margin
```
*inaccurate

DragShot replied:
Thank you for the heads-up.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/MaterialWindow.java:
```diff
@@ -8,45 +8,439 @@
             setLayout(new BorderLayout());
             setOpaque(false);
-            setBorder(new EmptyBorder(MaterialShadow.OFFSET_TOP, MaterialShadow.OFFSET_LEFT, MaterialShadow.OFFSET_BOTTOM, MaterialShadow.OFFSET_RIGHT));
+            setDeployed(true);
```
What does `deployed` mean in this context?

DragShot replied:
Oh, I see I forgot to add docs for that. When a `ShadowPane` is "deployed", it wraps the window contents, casting a shadow around them. This doesn't happen, for instance, when the window is maximized, so `setDeployed()` is called from `MaterialWindow.setExtendedState()` to adjust this value as necessary.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/Utils.java:
```diff
@@ -0,0 +1,185 @@
+    public static boolean isTranslucencySupported(){
+        boolean nativeTrans;
+        if (System.getProperty("java.version").contains("1.6")) {
+            System.err.println("Per-pixel translucency is currently not "
```
This is not really an error and we shouldn't write to `stderr` for this. Some Linux setups (e.g. mine) don't even support this with Java 8. :roll_eyes:

DragShot replied:
They don't? Wow that's weird. I wouldn't expect that from Linux.

Anyway, you said the minimum source-level is 1.7 right? Then I can safely remove it.

I did some tests in a spare Windows2K VM I had at work, thanks to it I figured out some issues there were with font antialiasing when it is disabled by default by your Desktop configuration. As it is a quite an old OS, I needed to compile to Java 6 in order to run my tests, and `GraphicsDevice.isWindowTranslucencySupported()` doesn't even exist in Java 6.

Besides, the code actually comes from [DS Desktop Notify](https://github.com/DragShot/DS-Desktop-Notify).

leMaik commented on src/main/java/de/craften/ui/swingmaterial/Utils.java:
```diff
@@ -0,0 +1,185 @@
+    /**
+     * Determines if a given {@link Color} is dark enough for white text to be
+     * seen more easily than black text. This tries to stick to the Material
+     * Color Guide as much as possible, and although two or three of the color
```
Instead of writing a "magic" formula, we could just hard-code that for the `MaterialColor`s and add a `isDark()` method to the enum.
Or at least, add that method to the enum, it might be useful. :wink:

DragShot replied:
Actually, that was going to be my initial approach, before an important question came out of nowhere:

**What if the user wants to use *their own* color?**

Time teaches you to expect the unexpected.

Besides, it isn't exactly an expensive operation, so...

leMaik replied:
Supporting _any_ color is great! But it should at least give the correct results for the Material colors (i.e. the colors from the specs).

leMaik commented on src/main/java/de/craften/ui/swingmaterial/fonts/MaterialIcons.java:
```diff
@@ -0,0 +1,976 @@
+ * @author DragShot
+ */
+public class MaterialIcons {
+    public static final char _3D_ROTATION = '\uE84D';
```
Maybe rename this to `ROTATION_3D`? :wink:

DragShot replied:
`3d rotation` is the canonical name of that icon. Changing it would be *"too confusing for new players"*.

Please, tell me you got it... God, so much time away from TF2 is really starting to mess with me.

leMaik replied:
:smile: 
When I'd search for this icon, I'd probably type "rotation 3d" and still find "rotation_3d" as well as "_rotation_3d". I just don't like the leading underscore.

DragShot replied:
Well then... why not both?

leMaik replied:
Because nope. :smile: 
Edit: Because it's pretty uncommon to start constants with an underscore. See the [Oracle Code Conventions](http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-135099.html#367).

leMaik commented on src/main/java/de/craften/ui/swingmaterial/fonts/MaterialIcons.java:
```diff
@@ -0,0 +1,976 @@
+        } catch (IOException | FontFormatException e) {
+            throw new RuntimeException("Could not load " + resourceName, e);
+        }
+        //JDK 6 compatible
```
We don't target JDK 6 so this can be dropped.

DragShot replied:
It was part of my testing in Windows2K, as there isn't try-with-resources in Java 6.

leMaik commented on src/main/java/de/craften/ui/swingmaterial/fonts/MaterialIcons.java:
```diff
@@ -0,0 +1,976 @@
+//        }
+    }
+    
+    /*public static void main(String[] args) throws Exception {
```
Ah nice, the generator code. :+1: Forget about my comment above.

DragShot replied:
It could be useful for updating the Material Icons set in the future.

leMaik replied:
Absolutely, we can keep this.

leMaik reviewed (changes requested):
95% good to merge. I can't wait for this! :tada:

DragShot commented on src/main/java/de/craften/ui/swingmaterial/fonts/MaterialIcons.java:
```diff
@@ -0,0 +1,976 @@
+//        }
+    }
+    
+    /*public static void main(String[] args) throws Exception {
```
Yeah, it was part of my testing in Windows2K, as there isn't try-with-resources in Java 6.

leMaik reviewed (changes requested):
99.99% ready for merge. :+1:

DragShot commented:
Due to the observations I'm making a last commit tonight. It won't affect the public API so it is safe to use the current state of my fork as reference, it will only modify internal calls.

leMaik commented:
@DragShot Please tell me when the commit is comitted, I'll merge this PR then.

DragShot committed:
Some docs have been modified.Changes:
- Now MaterialTextField, MaterialFormattedTextField, MaterialPasswordField and MaterialComboBox use the foreground color for the bottom line and hint when idle.
- Popups in MaterialComboBox now use the foreground and background color from their combobox.
- Added a method in Utils for applying an alpha mask into a Color.
- Minor adjustments in preparation for PR.
- Some docs have been added.

DragShot committed:
Background color is no longer painted over the parent container.

DragShot committed:
Some docs have been added.

DragShot commented:
@leMaik All ready.

leMaik commented:
@DragShot :tada:
